### PR TITLE
fix: Apply visual states on date change

### LINKED ISSUE #6256: DatePicker VisualStateGroup "HasDateStates" does not work in Uno
## Current behavior

In UWP, DatePicker gives us additional visual state that we can use depending on whether the user has selected a date or not.

```xaml
<VisualStateGroup x:Name="HasDateStates">
	<VisualState x:Name="HasDate">
		<VisualState.Setters>
			<Setter Target="PlaceholderText.Visibility"
				Value="Collapsed" />
			<Setter Target="DateText.Visibility"
				Value="Visible" />
			<Setter Target="RootGrid.Background"
				Value="Green" />
		</VisualState.Setters>
	</VisualState>
	<VisualState x:Name="HasNoDate">
		<VisualState.Setters>
			<Setter Target="PlaceholderText.Visibility"
				Value="Visible" />
			<Setter Target="DateText.Visibility"
				Value="Collapsed" />
			<Setter Target="RootGrid.Background"
				Value="Red" />
		</VisualState.Setters>
	</VisualState>
</VisualStateGroup>
```

However, when used with uno, it does not seems to work.
actual

## Expected behavior

expected

## How to reproduce it (as minimally and precisely as possible)

Download this sample [DatePickerTransition.zip](https://github.com/unoplatform/uno/files/6649958/DatePickerTransition.zip)
Deploy on any other platforms than UWP (to see the bug)
Select a date

## Workaround

Technically, manually change the visibility depending on `SelectedDate` property with a Converter works.

## Environment

Nuget Package:

- [x] Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia
- [ ] Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia
- [ ] Uno.SourceGenerationTasks
- [ ] Uno.UI.RemoteControl / Uno.WinUI.RemoteControl
- [ ] Other: 

Nuget Package Version(s): 

Affected platform(s): 3.9.0-dev.17

- [x] iOS
- [x] Android
- [x] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Skia
  - [ ] WPF
  - [ ] GTK (Linux)
  - [ ] Tizen
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

IDE:

- [ ] Visual Studio 2017 (version: )
- [x] Visual Studio 2019 (version:16.8.4)
- [ ] Visual Studio for Mac (version: )
- [ ] Rider Windows (version: )
- [ ] Rider macOS (version: )
- [ ] Visual Studio Code (version: )

Relevant plugins:

- [ ] Resharper (version: )

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #6256

## PR Type

What kind of change does this PR introduce?

- Bugfix

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Controls/DatePicker/DatePicker.cs [hunk 1]
```diff
@@ -1097,6 +1097,8 @@ async void ShowPickerFlyout()
 			 DateTimeOffset oldValue,
 			 DateTimeOffset newValue)
 		{
+			UpdateVisualState();
```
[the remark this answers is not in the document]
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/DatePicker/DatePicker.cs:1100]: @Youssef1313 no that's correct. But we should find a way to protect the feature against regression once #3519 will be implemented.

**Comment by Youssef1313** [comment]
Using `TestServices.WindowHelper.WindowContent = datePicker;` fixed part of my testing problem (Thanks to @MartinZikmund).
`GetVisualStateGroups` now doesn't return null. But it only returned CommonStates. It didn't return HasDateStates which I'm interested in tests

**Comment by Youssef1313** [comment]
NOTE: I cannot find anything related to DatePicker in https://github.com/unoplatform/uno/blob/master/src/Uno.UI/UI/Xaml/Style/Generic/Generic.xaml.

However, I can find for UWP in `C:\Program Files (x86)\Windows Kits\10\DesignTime\CommonConfiguration\Neutral\UAP\10.0.17763.0\Generic`.

This is probably somehow related.

**Comment by jeromelaban** [comment]
The themes for this control are located here: https://github.com/unoplatform/uno/blob/master/src/Uno.UI/UI/Xaml/Controls/DatePicker/DatePicker_themeresources.xaml

**Comment by Youssef1313** [comment]
These are not returned by `GetVisualStateGroups` as shown in the previous screenshot. Any idea why or what I'm doing wrong?

**Comment by jeromelaban** [comment]
> These are not returned by `GetVisualStateGroups` as shown in the previous screenshot. Any idea why or what I'm doing wrong?

No, but if the visual states are not returned, maybe it's not the right control or style?

**Comment by Youssef1313** [comment]
@jeromelaban `VisualTreeHelper.GetChild(datePicker, 0)` is returning a `StackPanel`, which seems to be coming from `Generic.xaml`, not the file you referenced:

https://github.com/unoplatform/uno/blob/215593e449f15cb9a1a8a4126a42763d5911ddd4/src/Uno.UI/UI/Xaml/Style/Generic/Generic.xaml#L7305-L7328

I got no idea why this is happening.

**Review by jeromelaban — APPROVED** [review]

**Comment by jeromelaban** [comment]
Let's merge it as-is for now, it's unlikely to have any incorrect effects.

**Pull request merged**